### ISSUE TITLE
Backcall operator

### ISSUE THREAD

**Issue by nythrox** [DanielXMoore/Civet#1558]
Can we get a backcall operator like in [Livescript ](https://livescript.net/#functions:~:text=Backcalls) and [Glean](https://tour.gleam.run/advanced-features/use/)?

```
do
  data <- $.get 'ajaxtest'
  $ '.result' .html data
  processed <- $.get 'ajaxprocess', data
  $ '.result' .append processed
  alert 'hi'
```
Gets converted into
```
$.get('ajaxtest', function(data){
  $('.result').html(data);
  $.get('ajaxprocess', data, function(processed){
    $('.result').append(processed);
  });
});
alert('hi');

```

This frees us from needing to ask the language developer to add features like await/async, try/catch, generators, etc... 
And gives functional programmers and framework developers a whole new level of power without ruining the ergonomy of the framework (see: effect-ts).

I'd hope for a syntax which is not as unnatural as gleam or scala, and more to [Roc](https://www.roc-lang.org/tutorial#the-!-suffix) and Idris which use the `!` suffix (allowing an easy suffix like `!` also gives us `?.` and any other type of chaining without devs having to implement it)

**Comment by bbrk24** [comment]
Isn't this just `util.promisify`? Even if you aren't in Node, this isn't hard to implement yourself:

```js
const promisify = fn => (...args) =>
  new Promise((resolve, reject) =>
    fn(...args, (result, error) => error == null ? resolve(result) : reject(error))
  );
```

**Comment by nythrox** [comment]
@bbrk24 
No, backcall operator isn't a higher order function. Its a syntactic feature which moves everything after it into a callback.
For example:
```
const msg <- Promise.resolve(0).then;
console.log(msg);
```
gets turned into
```
Promise.resolve(0).then((msg ) -> console.log(msg))
```

It can't be done in userland since it rearranges the order of statements in an expression, while maintaining the appearance of a normal control flow (direct-style).

Await/async does this too, but backcalls are more general, since they move everything after the backcall into a callback, being passed into the function on the right side of the backcall operator.

**Comment by bbrk24** [comment]
I'm still not sure I understand the point. Your initial motivating example still feels like it could just use `promisify`:
```civet
get := util.promisify $@get
async do
  data := await get 'ajaxtest'
  $('.result').html data
  processed := await get 'ajaxprocess', data
  $('.result').append processed
alert 'hi'
```
`async do` isolates the asynchronicity, meaning the outside scope doesn't need to wait for it to complete. The `alert 'hi'` at the bottom will be hit while the `await get 'ajaxtest'` is waiting.

Your second example,
> ```civet
> const msg <- Promise.resolve(0).then;
> console.log(msg);
> ```
feels to me like a convoluted way to avoid saying `await`.

As for the third example, I really don't see the point in saying `x <- (fn) -> fn(...)`. That seems like a really convoluted way of just assigning directly to `x`.

**Comment by edemaine** [comment]
I guess one interesting thing about backcalls, compared to `promisify`, is that in principle everything could remain synchronous in the JavaScript sense. For example, if `$.get` was synchronous and then called the callback, then everything would resolve synchronously. It's kind of like continuations...

That said, I'm not aware of many scenarios where callbacks are used in a synchronous fashion; they're mostly for asynchronous behavior. In that case, `promisify` + `await` seems simpler for anyone familiar with promises.

By the way, have you seen [IcedCoffeeScript](https://maxtaco.github.io/coffee-script/)? I used to use it, back before ES gained promises and its own notion of `await`, but meanwhile it offered a pretty nice `await`/`defer` syntax for what I think is backcalls:

```coffee
await $.getJSON url, defer json
console.log json
↓↓↓
$.getJSON url, (json) ->
  console.log json
```

It also supported loops and such: [[playground](https://maxtaco.github.io/coffee-script/#try:results%20%3D%20new%20Array(list.length)%0Aawait%0A%20%20for%20item%2C%20i%20in%20list%0A%20%20%20%20fetch%20item%2C%20defer%20results%5Bi%5D%0Aconsole.log%20results)]

```coffee
results = new Array(list.length)
await
  for item, i in list
    fetch item, defer results[i]
console.log results
```

I liked it at the time, and it took me a while to understand and transition to promises and `await`. But nowadays I'd prefer Civet's `console.log await.all fetch item`, which is equivalent to the above IcedCoffeeScript (in the `async` world). As bbrk24 points out, `async do console.log await.all fetch item` hides the `async` aspect; the only difference is if `fetch` was actually synchronous.

**Comment by bbrk24** [comment]
> `console.log await.all fetch item`

`console.log await.all list.map fetch .`?

**Comment by nythrox** [comment]
> I'm still not sure I understand the point. Your initial motivating example still feels like it could just use promisify:
This is a contrived example to show how to do callback-oriented code without await/async

>feels to me like a convoluted way to avoid saying await.
You are absolutely correct. This is exactly await/async, except that it doesn't only work for Promises, but any callback that has the structure of a promise (monadic structures). Which means it can work for exceptions, promises, nullables, iterators, generators, coroutines, promises, and many other control-flow structures that wouldn't need to be added to the language itself. This means we could add powerful features in user-land and make them usable with this simple sugar syntax, without having to ask language developers to implement custom syntax for every feature.

**Comment by nythrox** [comment]
@edemaine 
> It's kind of like continuations...

Yes, this lets you write code that uses continuations (callbacks) in direct-style, ie, in the sequential fashion that async/await allows.
Continuations are very powerful and a -lot- can be done on top of them, but the reason we don't see much usage of it is due to how ugly it is to use them (callback hell).

Backcall operator solves this for all constructs that can be built with continuations (iterators, async/await, streams, even request handlers) and would be a game changer for people who are actively building these constructs from scratch, like the Effect-TS guys, reactivex, old promise libraries, coroutine libraries, http libraries and etc. to have much more control and power over their code.

**Comment by nythrox** [comment]
the do notation is also universal in functional programming languages, which don't ever need to introduce custom syntax for things like await/async, exceptions, iterators, nullables, and more, since the do notation is sufficient for solving the legibility problem of monads (which encapsules almost all control flow constructs and much more)

**Comment by nythrox** [comment]
I feel like I am repeating myself here and I want to correctly transmit the WHY of do notation. Please help me out here if my explanations aren't what you guys are looking for.

To summarize:
- Do notation is await/async, but generalized for any construct (like promises) that can be built with callbacks
- Callbacks are very powerful, can and are used to build things like promises, coroutines, generators, iterators, exceptions, backtracing, deferring, nullables, streams and more.
- Working with callbacks sucks (because of callback hell), this sugar syntax fixes it for all cases above. Currently it is uncommon to work with anything but promises, but that is partially because of how hard it is to work with callbacks. Libraries like RxJs and Effect-TS are very powerful but ugly to use and read, and this would in large part be solved by allowing us to write code in direct style (like await async)

**Comment by edemaine** [comment]
We brainstormed notation for backcalls a bit on Discord. Before I forget, here was an idea:

* I don't like how the standard `<-` always adds the callback as a last argument.
* [IcedCoffeeScript](https://maxtaco.github.io/coffee-script/)'s `defer` fixed this problem, but it's obviously not something we can use as is (`await` is taken).
* Imagine that `foo <- bar` means `bar((foo) => rest of block`.
* Then the first example could be written as `data <- $.get 'ajaxtest', .`, using Civet's function partial notation `.`. This lets you specify where the callback goes in a general way.
* Another example is `foo <- (done) => a; b; c; done(d)`. Looks a bit like `new Promise`...

**Comment by ceremcem** [comment]
As a die-hard LiveScript user, I can say that backcalls make life a lot easier.

**Comment by edemaine** [comment]
> As a die-hard LiveScript user, I can say that backcalls make life a lot easier.

To help direct the design, we'd love to have more examples, either as links to existing code bases or pasted here.

**Comment by ceremcem** [comment]
It's like using one-liner sync function call and variable assignments in Javascript. Eg: 

https://github.com/aktos-io/scada.js/blob/6f0a53879052e1b8e1cd3464cba82b051f254996/components/export-to-csv/index.ls#L16

And if you introduce backcalls with bound functions (`<~`), it's even more useful: 

https://github.com/aktos-io/aktos-dcs-node/blob/180789c4ade01c41cd3b9e011d56a2e38a838264/src/auth-request.ls#L45

Please see more of my use cases: 
1. https://github.com/search?q=repo%3Aaktos-io%2Faktos-dcs-node+%3C%7E&type=code
2. https://github.com/search?q=repo%3Aaktos-io%2Fscada.js+%3C~&type=code
3. https://github.com/search?q=repo%3Aaktos-io%2Faecad+%3C~&type=code

Best use case of backcalls is that you don't have to indent everything if you decide to insert a function call in the middle of your code later (for example, you want to wait for server response and `await` is not an option).

**Comment by vendethiel** [comment]
> I don't like how the standard `<-` always adds the callback as a last argument.

fwiw, that's not even true in livescript, you get a `_` placeholder:

```lsc
a <- b _, c
d
```

**Comment by caasi** [comment]
+1 for this feature. @nythrox++

I come across this issue when looking for a livescript replacement.

When livescript introduced backcall, async/await wasn't even a thing. And backcall solves CPS-transformation in such a simple way. It also shows that callback hell is an illusion if you can just change the way you indent your code (which is fun but hard to explain to new comers).

Limited solutions like async/await and the use hook in React make scripting less fun in my opinion.

BTW do notation is great but not essential. Paul Hudak uses list comprehension to archive the same thing in The Haskell School of Expression.

Maybe Civet can come up a different solution with the same spirit inside.

**Comment by STRd6** [comment]
I'm rereading this and am in favor of adding back-calls largely compatible with LiveScript.

One change: instead of `_` for the argument placeholder we'll use Civet's `.` so we don't shadow the potential use of `_` as a variable. We may consider a flag for LiveScript compat support of `_` as a back-call placeholder if [people are interested in more LiveScript compat](https://github.com/DanielXMoore/Civet/issues/1729).

I've thought a little bit more on the syntax and can't really come up with anything substantially better. I've grown more accepting of inserting an invisible/implied final argument as the callback in cases where there is no explicit placeholder. It makes sense for general continuation passing.

Example:
```
data: DataType <- $.get 'ajaxtest'
// rest of block
---
$.get('ajaxtest', function(data: DataType){
// rest of block
})
```

As for assigning / declaring I think using the declaration in the function and having it naturally shadow anything prior in the block. Also note how we can fit type annotations into the function parameter declarations.

We should also keep `<~` as the syntax for bound back-calls corresponding to `=>`. Ideally it would have been `<=` but that's already taken by the mathematical comparison. A case could be also be made for `<:`.


I think the implementation is fairly straightforward but building up some good test cases to iron out any edges would be useful.

**Comment by edemaine** [comment]
Cool! I just reread the thread as well. Just so I understand the proposal: the notation is `lhs <- rhs` where

1. `lhs` can be `identifier` or `identifier: type` or presumably `pattern` like `[a, b]` or `[a:: aType, b:: bType]`; basically any function parameter.
   * We should probably also allow `lhs` to be empty, as [LiveScript](https://livescript.net/#functions-backcalls) does, to support callbacks without arguments (when there's no result of interest).
2. `rhs` must be a function call (at least for now). (a) If it has a `.` placeholder, the callback gets added there. (b) Otherwise, it gets added as a final argument.

Is the constructed function a `->` (`function`) or `=>` function? Your code example uses `->` but later on you say it should be `=>`. I imagine `=>` is the case you normally want; it'd probably be jarring to have `this` rebound, and I don't see a setting where you'd want to assign `this` (the notation doesn't really allow for it). Would we ever want a generator inside a callback? We could use `wrapIIFE`'s logic to use `=>` or `function*` as appropriate (in the latter case, passing in `this` to preserve it).

Notationally, I don't like that `<-` usually generates `=>`... We actually use `<:` for `implements`, so that might be confusing. Another option is `<==`. It looks like [LiveScript](https://livescript.net/#functions-backcalls) uses `<-` for `->` functions, and `<~` for `~>` functions which are similar to `=>`. (Incidentally, `!=>` is LiveScript's notation to skip implicit return; interesting.) This has a nice symmetry, but I'm not sure whether `function` is actually useful here... I think LiveScript did it because ES `=>` didn't exist yet,.

An alternative design (mentioned above) is that `rhs` can be any expression, and it gets called with a single argument. This feels more symmetric to how pipelines work. It's actually equivalent in the `.` placeholder case (a). We could make it fully equivalent to the current proposal by adding an exception that, if `rhs` is a function call without placeholders, then it gets a callback added as final argument. This feels pretty irregular though, and asymmetric with pipelines... Maybe this explains why I'm slightly uncomfortable with part (b) of the proposal.

But running with the proposal, we could in the future define what `lhs <- rhs` means when `rhs` is not a function call. For example, we could try to define it for an array of function calls (do various things and wait for them all to complete, like `await.all` or my IcedCoffeeScript example):

```js
results <-
  for x of y
    f x
g results
---
const results = [], callback = () => {
  g(results)
}
let done = 0, total = 0, ready = false
for (x of y) {
  const i = total++
  f(x, (result) => {
    results[i] = result
    if (ready && done >= total) callback()
  })
}
ready = true
if (done >= total) callback()
```

Still not as powerful as IcedCoffeeScript's `defer`, but maybe covers some useful cases, similar to `await` operators...

Back to pipelines, `<-` might be useful in pipelines.

```js
data
|> fetch
|> preprocess
|> <- render
|> postprocess
```

Except I don't know what that `render` call should look like. `render(lastPipeResult, callback)`? At this point we might need two placeholders, one for where the pipeline argument goes and one for where the callback goes. It might be worth looking for a notation that makes pipelines possible. Though I guess LiveScript didn't solve this problem, despite having both features.

(Incidentally, just bumped into [this old CoffeeScript thread](https://github.com/jashkenas/coffeescript/issues/4957) on the same topic. One cute notation idea is `rhs => \` to mean "function is continued on future lines without indentation". I don't think it's as nice as `<-` though.)

**Comment by STRd6** [comment]
The `lhs` can also be multiple comma separated function arguments so I think it's actually an arguments list with implicit parentheses. This should also cover the zero item list as well.

Agree on `rhs` must be a function call with optional placeholder and an implied placeholder as the final argument if not present. The function call may be explicit or implicit with parens.

If we find a good pattern for handling the pipeline back-call combo we could get pretty cute combining the arrows and pipes into one inscrutable symbol :P

**Comment by ceremcem** [comment]
> Back to pipelines, <- might be useful in pipelines.
>
> ```
> data
> |> fetch
> |> preprocess
> |> <- render
> |> postprocess
> ```

Now you have gone too far XD

**Comment by vendethiel** [comment]
> Incidentally, `!=>` is LiveScript's notation to skip implicit return; interesting

[The reason is a 12+ years old pun on "not function"](<https://github.com/satyr/coco/issues/207#issuecomment-14747142>) 

> 1. We should probably also allow `lhs` to be empty, as [LiveScript](https://livescript.net/#functions-backcalls) does, to support callbacks without arguments (when there's no result of interest).

You need to think of a syntax to annotate the return type of such a backcall I think, and that's not going to be pretty

**Comment by STRd6** [comment]
> Is the constructed function a `->` (`function`) or `=>` function? Your code example uses `->` but later on you say it should be `=>`. I imagine `=>` is the case you normally want; it'd probably be jarring to have `this` rebound, and I don't see a setting where you'd want to assign `this` (the notation doesn't really allow for it). Would we ever want a generator inside a callback? We could use `wrapIIFE`'s logic to use `=>` or `function*` as appropriate (in the latter case, passing in `this` to preserve it).
> 
> Notationally, I don't like that `<-` usually generates `=>`... We actually use `<:` for `implements`, so that might be confusing. Another option is `<==`. It looks like [LiveScript](https://livescript.net/#functions-backcalls) uses `<-` for `->` functions, and `<~` for `~>` functions which are similar to `=>`. (Incidentally, `!=>` is LiveScript's notation to skip implicit return; interesting.) This has a nice symmetry, but I'm not sure whether `function` is actually useful here... I think LiveScript did it because ES `=>` didn't exist yet,.

I meant in my example for `<-` to translate to `function` and `<~` to translate to `=>`. What I meant by "ideally we'd use `<=`" was instead of `<~` but we can't because it collides with the mathematical operator. Good point about `<:` so I think using `<~` for `=>` is the way to go.
